sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -614,47 +742,103 @@ checkFields :: UpgradedRecordOrigin -> Upgrading [(FieldName, Type)] -> TcUpgrad
+    different <- filterHashMapM (fmap not . isUpgradedType) existing
     when (not (HMS.null different)) $
-        throwWithContextF present (EUpgradeRecordFieldsExistingChanged origin)
+        throwWithContextF present' (EUpgradeRecordFieldsExistingChanged origin)
     when (not (all newFieldOptionalType new)) $
         case origin of
           VariantConstructor{} ->
-            throwWithContextF present (EUpgradeVariantAddedVariantField origin)
+            throwWithContextF present' (EUpgradeVariantAddedVariantField origin)
           _ ->
-            throwWithContextF present (EUpgradeRecordFieldsNewNonOptional origin)
+            throwWithContextF present' (EUpgradeRecordFieldsNewNonOptional origin)
         -- If a new field has a non-optional type
     -- If the order of fields changed
     when (not $ and $ foldU (zipWith (==)) $ fmap (map fst) fields) $
-        throwWithContextF present (EUpgradeRecordFieldsOrderChanged origin)
+        throwWithContextF present' (EUpgradeRecordFieldsOrderChanged origin)
     where
         newFieldOptionalType (TOptional _) = True
         newFieldOptionalType _ = False
 
 -- Check type upgradability
-checkUpgradeType :: SomeErrorOrWarning e => Upgrading Type -> e -> TcUpgradeM ()
-checkUpgradeType type_ err = do
-    sameType <- isSameType type_
-    unless sameType $ diagnosticWithContextF present err
+checkType :: SomeErrorOrWarning e => Upgrading Type -> e -> TcUpgradeM ()
+checkType type_ err = do
+    sameType <- isUpgradedType type_
+    unless sameType $ diagnosticWithContextF present' err
 
-isSameType :: Upgrading Type -> TcUpgradeM Bool
-isSameType type_ = do
+isUpgradedType :: Upgrading Type -> TcUpgradeM Bool
+isUpgradedType type_ = do
     expandedTypes <- runGammaUnderUpgrades (expandTypeSynonyms <$> type_)
-    let strippedIdentifiers = fmap unifyTypes expandedTypes
-    pure (foldU alphaType strippedIdentifiers)
-
-unifyIdentifier :: Qualified a -> Qualified a
-unifyIdentifier q = q { qualPackage = PRSelf }
-
-unifyTypes :: Type -> Type
-unifyTypes typ =
-    case typ of
-      TNat n -> TNat n
-      TSynApp n args -> TSynApp (unifyIdentifier n) (map unifyTypes args)
-      TVar n -> TVar n
-      TCon con -> TCon (unifyIdentifier con)
-      TBuiltin bt -> TBuiltin bt
-      TApp fun arg -> TApp (unifyTypes fun) (unifyTypes arg)
-      TForall v body -> TForall v (unifyTypes body)
-      TStruct fields -> TStruct ((map . fmap) unifyTypes fields)
-
+    deps <- view upgradingDeps
+    let checkSelf = alphaTypeCon
+        checkImport pastT presentT =
+          case (qualPackage pastT, qualPackage presentT) of
+            (PRImport pastPkgId, PRImport presentPkgId) ->
+              pastPkgId == presentPkgId ||
+              (case (pastPkgId `HMS.lookup` deps, presentPkgId `HMS.lookup` deps) of
+                (Just (pastName, pastVersion), Just (presentName, presentVersion)) -> pastName == presentName && pastVersion < presentVersion
```

Comment:
Note that in the participant check this works because we've treated the deps in topological order, so we know that "p1.version < p2.version" is a proxy for "p2 upgrades p1". I'm not sure we tread them in any particular order here.

Reply:
Dependencies are in topological order, interestingly enough - the dependency list in an LF package is set via GetDependencies, which always returns items in topological order.

Reply:
This was a lie - dependencies are not in topological order, only modules from GetDependencies are in the correct order